sdk/evm/execution_error.go

```
@@ -0,0 +1,931 @@
+	if strings.Contains(errStr, revertPrefix) {
+		// Try to extract the plain string revert reason
+		// Format: "execution reverted: revert: <reason>" or "revert: <reason>"
+		revertIdx := strings.Index(errStr, revertPrefix)
+		if revertIdx != -1 {
+			// Extract everything after "revert: "
+			reason := strings.TrimSpace(errStr[revertIdx+len(revertPrefix):])
+			if reason != "" {
+				return revertReasonData{
+					Decoded: reason,
+				}
+			}
+		}
+	}
+
+	// Try to extract from "custom error 0x...: <hex data>" format
+	if strings.Contains(errStr, "custom error") {
+		customErr := extractCustomErrorRevertData(errStr)
+		if customErr != nil {
+			rawData := customErr.Combined()
+			decoded := decodeRevertReasonFromCustomError(customErr)
+
+			return revertReasonData{
+				RawData:     rawData,
+				Decoded:     decoded,
+				CustomError: customErr, // Store CustomErrorData directly for easy selector access
+			}
+		}
+	}
+
+	// Try to extract revert data from error string using array, hex or bytesArray formats
+	if strings.Contains(errStr, "CallReverted") {
+		if rawData := extractBytesArrayRevertData(errStr); len(rawData) > 0 {
+			return revertReasonData{
+				RawData: rawData,
+				Decoded: decodeRevertReason(rawData),
+			}
+		}
+	}
+
+	if rawData := extractHexEncodedRevertData(errStr); len(rawData) > 0 {
+		return revertReasonData{
+			RawData: rawData,
+			Decoded: decodeRevertReason(rawData),
+		}
+	}
+
+	if rawData := extractBytesArrayRevertData(errStr); len(rawData) > 0 {
+		return revertReasonData{
+			RawData: rawData,
+			Decoded: decodeRevertReason(rawData),
+		}
+	}
+
+	return revertReasonData{}
+}
+
+// parseBytesFromString parses a string representation of bytes like "8 195 121 160 ..."
+func parseBytesFromString(s string) []byte {
+	parts := strings.Fields(s)
+	if len(parts) == 0 {
+		return nil
+	}
+	bytes := make([]byte, 0, len(parts))
+	for _, part := range parts {
+		if val, err := strconv.ParseUint(part, 10, 8); err == nil {
```

we're silently ignoring paring errors here. Isn't it a bit risky?

---

Since this is just an attempt to add structure to an RPC error, I don't think we'd want to hijack the actual rpc error and add noise because of the decoding process. I opted for just ignoring this which would just endup returning the original error from the rpc call.